ftrack_sftp_accessor/sftp.py

```diff
@@ -0,0 +1,256 @@
+        )
+
+    @property
+    def ssh(self):
+        """Return SSH resource."""
+        if self._ssh is None:
+            self._logger.debug("Initialising SSH Session")
+            self._ssh = paramiko.SSHClient()
+            self._ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+            if not self._password:
+                # attempt to connect using keyfile if no password specified
+                self._ssh.connect(
+                    self._host,
+                    port=self._port,
+                )
+            else:
+                self._ssh.connect(
+                    hostname=self._host,
+                    port=self._port,
+                    username=self._username,
+                    password=self._password,
+                )
+        return self._ssh
+
+    @property
+    def sftp(self):
+        """Return SFTP resource."""
+        self._logger.debug("Initialising SFTP Session")
+        if self._sftp is None:
+            self._sftp = self.ssh.open_sftp()
+
+        return self._sftp
+
+    def list(self, resource_identifier):
+        """Return list of entries in *resource_identifier* container.
+
+        Each entry in the returned list should be a valid resource identifier.
+
+        Raise :py:class:`~ftrack.ftrackerror.AccessorResourceNotFoundError` if
+        *resourceIdentifier* does not exist or
+        :py:class:`~ftrack.ftrackerror.AccessorResourceInvalidError` if
+        *resourceIdentifier* is not a container.
+
+        """
+        if not resource_identifier.endswith("/"):
+            resource_identifier += "/"
+
+        if resource_identifier == "/":
+            resource_identifier = ""
+
+        try:
+            self.sftp.stat(resource_identifier)
+            return self.sftp.listdir(resource_identifier)
+        except IOError:
+            self._logger.debug(
+                f"Returning an empty list as resource identifier {resource_identifier} doesn't exist"
+            )
+            return []
+
+    def exists(self, resource_identifier):
+        """Return if *resourceIdentifier* is valid and exists in location."""
+        # Root directory always exists
+        if not resource_identifier:
+            return True
+
+        return self.is_container(resource_identifier) or self.is_file(
+            resource_identifier
+        )
+
+    def is_file(self, resource_identifier):
+        """Return whether *resource_identifier* refers to a file."""
+        # Root is a directory
+        if not resource_identifier:
+            return False
+
+        resource_identifier = resource_identifier.rstrip("/")
+
+        try:
+            file_object = self.sftp.stat(resource_identifier)
+        except IOError:
+            self._logger.debug(
+                f"Returning is not file as resource identifier {resource_identifier} doesn't exist"
+            )
+            file_object = None
+        except Exception as error:
+            raise AccessorOperationFailedError(
```

Comment:
@iwootten AccessorOperationFailedError has a different signature:

https://bitbucket.org/ftrack/ftrack-python-api/src/88b364e0731db41eebbbb164e5c2282c8c71e6cf/source/ftrack_api/exception.py#lines-295

Reply:
please review also the other exceptions raised in the code.

Reply:
@iwootten the AccessorOperationFailedError have no **details** argument, I believe you want to you the **error** one.